### ISSUE TITLE
Replicate Result

### ISSUE THREAD

**Issue by Kira0096** [lugiavn/gt-crossview#2]
Hello, I'm trying to replicate your result (with the model you share), but I couldn't seem to get it right. I don't know which part of my code is wrong. So I was wondering if you can share the code that produced the result you listed on paper and readme.

Thank you very much!

**Comment by lugiavn** [comment]
Hi, sorry for the trouble, but my Matlab (yes it's Matlab) testing script is really messy. I'm putting the outline here but you might need to spend some effort to make it work

**Comment by lugiavn** [comment]
First load the model in matlab, making sure the paths are all correct.
I separated 2 branches of the Siamese network into deploy model 1 and deploy model 2


```

clear;
save_path = 'saves/test_ranking_rotationinvariant/rankingdbl0_denver10kgt';


deploy_model1 = '/deploy1.prototxt';
deploy_model2 = 'deploy2.prototxt';
model_weights = '2_iter_80000.caffemodel';




streetview_txt = 'data/denver.streetview.txt';
overhead_txt = 'data/denver.overhead.txt';
data_root = 'data/';

caffe.set_mode_cpu();
caffe.set_mode_gpu();
caffe.set_device(0);
disp 'load model...'
%net = caffe.Net(deploy_model, model_weights, 'test');
net1 = caffe.Net(deploy_model1, model_weights, 'test');
net2 = caffe.Net(deploy_model2, model_weights, 'test');

disp 'load data...';
data = load_data_from_txt(streetview_txt, overhead_txt);

```


```

function data = load_data_from_txt( streetview_txt, overhead_txt , num, do_parse_line)
%LOAD_DATA_FROM_TXT Summary of this function goes here
%   Detailed explanation goes here

if nargin < 3
    num = 9999999999;
end
if num < 0
    num = 9999999999;
end
if nargin < 4
    do_parse_line = 1;
end

data = struct;
i = 0;
streetview_txt = fopen(streetview_txt, 'rt');
overhead_txt = fopen(overhead_txt, 'rt');

for zzz=1:num
    l1 = fgetl(streetview_txt);
    l2 = fgetl(overhead_txt);
    if length(l1) < 5
        break;
    end
    if do_parse_line
        l1 = strsplit(l1);
        l2 = strsplit(l2);
    else
        l1 = {l1, '-1'};
        l2 = {l2, '-1'};
    end
    
    i = i + 1;
    
    label = -1;
    if strcmp('1', l1{2})
        label = 1;
    end
    if strcmp('0', l1{2})
        label = 0;
    end
    
    data(i).label = label;
    data(i).streetview_img = l1{1};
    data(i).overhead_img = l2{1};
end

fclose(streetview_txt);
fclose(overhead_txt);

end

```

**Comment by lugiavn** [comment]
Then this compute all the image features:
```


rotation_test = [1 4 16];
use_gt_aligment = 0;

if (use_gt_aligment)
    rotation_test = 1;
end

rotation_num = max(rotation_test);

disp 'Compute feature...';

for i=1:length(data)
    if mod(i, floor(length(data)/100)) == 1,
        disp(['progress: ', num2str(i), '/', num2str(length(data))]);
    end
    
    % read image
    d = data(i);
    streetview_img = imread([data_root '/' d.streetview_img]);
    overhead_img = imread([data_root '/' d.overhead_img]);
    [a, b] = regexp(d.overhead_img, '.a[-0-9]+.o');
    north_angle = str2double(d.overhead_img(a(end)+2:b(end)-2));
    
    % compute feature for street view
    net1.blobs('data_streetview').set_data(convert_img_for_cafffe(streetview_img));
    net1.forward_prefilled();
    data(i).streetview_f = net1.blobs('pre1_fc8').get_data();

    % compute feature for overhead
    % do rotation & crop
    for j=1:rotation_num
        angle = (j - 1) * 360 / rotation_num - north_angle;
        if use_gt_aligment, angle = 0; end;
        rotated_overhead = imrotate(overhead_img, angle, 'nearest', 'crop');
        %rotated_overhead = rotated_overhead(150:end-150, 150:end-150,:);
        rotated_overhead = rotated_overhead(60:end-60, 60:end-60,:);
        
        % compute image feature
        net2.blobs('data_overhead').set_data(convert_img_for_cafffe(rotated_overhead));
        net2.forward_prefilled();

        % save
        data(i).overhead_f{j} = net2.blobs('pre2_fc8').get_data();
    end
end
```

```

function img = convert_img_for_cafffe(img, s)
%CONVERT_IMG_FOR_CAFFFE Summary of this function goes here
%   Detailed explanation goes here
    if nargin < 2
      s = 224;
    end
    img = single(imresize(img, [s s])) - 100.0;
    if size(img,3) == 1,
        img(:,:,2) = img(:,:,1);
        img(:,:,3) = img(:,:,1);
    end
    img = img(:, :, [3, 2, 1]); % convert from RGB to BGR
    img = permute(img, [2, 1, 3]); % permute width and height
end


```

**Comment by lugiavn** [comment]
Then compute the ranking, you'll need FLANN for Matlab

```
%% build kdd-trees
%  and perform nearest neighbors search for each streetview
params.algorithm = 'kdtree';
params.trees = 8;
params.checks = 64;

overhead_data = [data.overhead_f]';
overhead_data = [overhead_data{:}];

% for i=1:length(data), data(i).overhead_f = {}; end;

kdts = {};
idss = {};
distances = {};
for j=1:rotation_num
    kdts{j} = KDTreeSearcher(overhead_data(:, j:rotation_num:end)');
    tic; [idss{j}, distances{j}] = knnsearch(kdts{j}, [data(:).streetview_f]', 'K', round(length(data)/9)); toc;
    %tic; [idss{j}, distances{j}] = flann_search(overhead_data(:, j:rotation_num:end), [data(:).streetview_f], round(length(data)/9), params); toc;
    %idss{j} = idss{j}'; distances{j} = distances{j}'; 
end

clearvars overhead_data

%% ok compute rankings

truthsranks = {};

% test for each case of crop num
for r=rotation_test

    truthsrank = [];
    for i=1:length(data)

        % accumate all nearest neighbors and sort
        id = [];
        dis = [];
        for j=1:(rotation_num/r):rotation_num
            id = [id, idss{j}(i,:)];
            dis = [dis, distances{j}(i,:)];
        end
        [a, b] = sort(dis);
        id = id(b);
        id = unique(id, 'stable');

        % find rank
        rank = find(id == i);
        if isempty(rank)
            truthsrank(i) = length(data);
        else
            truthsrank(i) = rank;
        end
    end
    truthsrank = truthsrank / length(data);
    truthsrank = sort(truthsrank);

    truthsranks{end+1} = truthsrank;
end

save([save_path, 'data']);

%% viz
close all
plot([truthsranks{1} 1], [[1:length(data)] / length(data) 1], '--b');
hold on;
plot([truthsranks{2} 1], [[1:length(data)] / length(data) 1], '--r');
plot([truthsranks{3} 1], [[1:length(data)] / length(data) 1], '--g');
xlim([0 0.1]);
grid on
```

**Comment by Kira0096** [comment]
Thanks a lot, I think this solves the issue !

**Comment by OnePunchMann** [comment]
Hello, I try to use the dataSet for localization. However, I really don't know  how to map coordinates to Google Maps.  I can't figure out what's the x, y mean.

**Comment by Kira0096** [comment]
I realized that the images in test sets (denver/detroit/seattle) are aligned differently from that in training set. 
I wonder what I can do in order to align them? (may be rotation or crop ?)

**Comment by lugiavn** [comment]
Hi, we didn't keep the Google Map coordinates

**Comment by lugiavn** [comment]
Hi, test set is not aligned on purpose.
Training data: we tried to align the overhead image
Test data: overhead image is always north top, west left, east right, south bottom. We assume the street view image orientation is unknown, so aligning is not possible. We do orientation invariant training to tackle that, more info is in the paper.

**Comment by Kira0096** [comment]
ok, I got it, thx!

**Comment by Kira0096** [comment]
I know it's a little unrelated, but I wonder how do u 'tried to align the overhead image'. I couldn't find this part in your paper.

**Comment by lugiavn** [comment]
The Google streetview image is a 360 panorama with orientation, we make a crop and infer the "camera orientation", for example if it is pointing east, then we rotate the overhead so that top is east, bottom is west

**Comment by Kira0096** [comment]
ok, I see. I really appreciate your help. Thx!
